incrementalprune.go

```diff
@@ -0,0 +1,348 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package txn
+
+import (
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/lru"
+	"github.com/kr/pretty"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+const pruneTxnBatchSize = 1000
+const queryDocBatchSize = 100
+const pruneDocCacheSize = 10000
+
+// IncrementalPruner reads the transzaction table incrementally, seeing if it can remove the current set of transactions,
+// and then moves on to newer transactions. It only thinks about 1k txns at a time, because that is the batch size that
+// can be deleted. Instead, it caches documents that it has seen.
+type IncrementalPruner struct {
+	docCache     *lru.LRU
+	missingCache *lru.LRU
+	stats        PrunerStats
+}
+
+// PrunerStats collects statistics about how the prune progressed
+type PrunerStats struct {
+	DocCacheHits       int
+	DocCacheMisses     int
+	DocMissingCacheHit int
+	DocsMissing        int
+	CollectionQueries  int
+	DocReads           int
+	DocStillMissing    int
+	StashQueries       int
+	StashReads         int
+	DocQueuesCleaned   int
+	DocTokensCleaned   int
+	DocsAlreadyClean   int
+	TxnsRemoved        int
+	TxnsNotRemoved     int
+}
```

Comment:
More of a question rather than a comment: is there a preference for `int`, rather than say `uint` in the Juju code base? As a bikeshedding remark, none of those fields seem like that they should need to be able to consider negative values.

Reply:
len() returns an 'int', so for fields that aggregate that sort of data, it made sense to just put it into an int. At one point I explicitly used a uint64, but it turns out that kr/pretty encodes unit64 as a Hex value, which makes it hard to get nice debugging results.

I agree that uint would be appropriate, but given things like len() return int it often makes sense to just use the normal type.